crates/core/database/src/models/messages/model.rs

```
@@ -302,19 +361,97 @@ impl Message {
                     if let Some(mention) = capture.get(1) {
                         mentions.insert(mention.as_str().to_string());
                     }
                 }
+
+                // Second step of mass mention resolution.
+                for capture in RE_ROLE_MENTION.captures_iter(content) {
+                    let raw = capture.get(0).expect("No capture?").as_str();
+                    if raw == "@everyone" {
+                        mentions_everyone = true;
+                        // Can't break early here since we want to capture any role ids
+                    } else if raw == "@online" {
+                        mentions_online = true;
+                    } else if let Some(role_match) = capture.get(1) {
+                        role_mentions.insert(role_match.as_str().to_string());
+                    }
+                }
+
+                if !role_mentions.is_empty() {
+                    let server_data = db
+                        .fetch_server(server_id.unwrap().as_str())
+                        .await
+                        .expect("Failed to fetch server");
+
+                    role_mentions = role_mentions
+                        .iter()
+                        .filter(|role_id| server_data.roles.contains_key(*role_id))
+                        .cloned()
+                        .collect();
+
+                    role_model_mentions.extend(
+                        role_mentions
+                            .iter()
+                            .map(|role_id| server_data.roles.get(role_id).to_owned()),
+                    );
+                }
+            }
+        }
+
+        // Validate the user can perform a mass mention
+        if !config.features.mass_mentions_enabled
+            && (mentions_everyone || mentions_online || !role_mentions.is_empty())
+        {
+            return Err(create_error!(FeatureDisabled {
+                feature: "features.mass_mentions_enabled".to_string()
+            }));
```

Makes more sense to silently fail here

---

NB. especially if we want to gradually rollout the feature